ocaml/xapi/xapi_host_driver.ml

```diff
@@ -85,6 +87,160 @@ let reload_driver_selection ~__context ~host =
   with Forkhelpers.Spawn_internal_error (err, _, _) ->
     internal_error "Failed to reload driver selection (%s): %s" __FUNCTION__ err
 
+(* Versions are compared in segments separated by punctuation characters.
+   If both segments are integers, integer comparison is used, if at least
+   one of the segments is not an integer, lexicographic comparison is used.
+   Any segment compares larger against an empty segment.
+   Thus, "1.2.3" > "1.2";
+         "1.2.3" > "1.2."
+         "1.2.3" < "1.2.fix"
+         "1.2.3" < "1.2.3fix"
+*)
+module Version = struct
+  let compare a b =
+    let split str = Str.(split_delim (regexp "[.,:;+_-]") str) in
+
+    List.compare
+      (fun a b ->
+        Stdlib.(
+          match (int_of_string_opt a, int_of_string_opt b) with
+          | Some a, Some b ->
+              Int.compare a b
+          | _, _ ->
+              String.compare a b
+        )
+      )
+      (split a) (split b)
+end
+
+let get_driver_versions driver_name selected_dir all_versions_dir =
+  let selected_version =
+    try
+      let driver_realpath =
+        Unix.readlink (selected_dir // (driver_name ^ ".ko"))
+      in
+      Filename.(basename (dirname driver_realpath))
+    with _ ->
+      D.debug "Driver '%s' has no selected version" driver_name ;
+      ""
+  in
+  let active_version =
+    try
+      Result.value ~default:""
+        (Xapi_host_driver_helpers.get_version
+           ("/sys/module" // driver_name // "notes/note.xenserver")
+        )
+    with _ ->
+      D.debug "Driver '%s' has no active version" driver_name ;
+      ""
+  in
+  let driver_versions =
+    let arr = Sys.readdir (all_versions_dir // driver_name) in
+    Array.sort Version.compare arr ;
+    Array.to_list arr
+  in
+  (selected_version, active_version, driver_versions)
+
+(* Scans the multi-version drivers filesystem hierarchy, updating the drivers
+   already in the DB if necessary, removing the ones that have been deleted
+   from the filesystem, and adding new ones *)
+let discover ~__context ~host =
+  try
+    let driver_paths = get_drivers_base_path ~__context ~host in
+    let db_drivers_set, db_drivers_map =
+      Db.Host_driver.get_all_records ~__context
```

Comment:
You could use `get_all_records_where` matching on the `host` (and drop the if-statement below) to restrict the amount of data being transferred from the coordinator.

Reply:
I couldn't really grok `get_all_records_where` so decided to use this version, turns out this is an old problem: https://lists.xenproject.org/archives/html/xen-api/2013-02/msg00068.html

Reply:
Yes, it's odd, and best to look for examples in the code base.